subsys/bootloader/bl_validation/bl_validation.c

```diff
@@ -267,20 +331,32 @@ static bool validate_firmware(u32_t fw_dst_address, u32_t fw_src_address,
 		return false;
 	}
 
-	if (!(((u32_t)fwinfo >= fw_src_address)
-		&& (((u32_t)fwinfo + fwinfo->total_size)
-			< (fw_src_address + fwinfo->size)))) {
+	if ((fwinfo->size > (PM_S0_SIZE))
+		|| (fwinfo->size > (PM_S1_SIZE))
+		|| (fwinfo->total_size > fwinfo->size)) {
+		PRINT("Invalid size or total_size in firmware info.\n\r");
+		return false;
+	}
+
+	if (!region_within((u32_t)fwinfo, (u32_t)fwinfo + fwinfo->total_size,
```

Comment:
Please consider using helper variables for these sums, it would improve readability IMO.